src/x86_64/x86_64.h

```diff
@@ -66,9 +68,94 @@ static inline void disable_interrupts()
 
 typedef u64 *context;
 
-extern context running_frame;
+context allocate_frame(heap h);
+
+typedef struct cpuinfo {
+    /* For accessing cpuinfo via %gs:0; must be first */
+    void * self;
+
+    /* This points to the frame of the current, running context. Entry
+       points expect this to be the second (+8) field here. */
+    context running_frame;
+
+    /* syscall_enter switches to this stack before calling syscall. It
+       must be the third field (+16). */
+    void * syscall_stack;
+
+    /* common_handler switches to this stack when calling process_bhqueue */
+    void * bh_stack;
+
+    /* The default frame for when we're not in a thread or bh context. */
+    context misc_frame;
+
+    /* for bh processing, where page faults must be supported */
+    context bh_frame;
+
+    u32 id;
+    boolean in_bh;              /* temporary, probably bh will go away? */
+    boolean in_int;             /* to catch exceptions during int processing */
+    volatile boolean online;
+    volatile boolean ipi_wakeup;
 
-extern void * syscall_stack_top;
+    /* The following fields are used rarely or only on initialization. */
+
+    /* stack for page faults, switched by hardware */
+    void * fault_stack;
+
+    /* stack for int handlers, switched by hardware */
+    void * int_stack;
+} *cpuinfo;
+
+extern struct cpuinfo cpuinfos[];
+
+static inline void cpu_setgs(int cpu)
+{
+    u64 addr = u64_from_pointer(&cpuinfos[cpu]);
+    write_msr(KERNEL_GS_MSR, 0); /* clear user GS */
+    write_msr(GS_MSR, addr);
+}
+
+static inline cpuinfo current_cpu(void)
+{
+    u64 addr;
+    asm volatile("movq %%gs:0, %0":"=r"(addr));
+    return (cpuinfo)pointer_from_u64(addr);
+}
+
+static inline context get_running_frame(void)
+{
+    return current_cpu()->running_frame;
+}
+
+static inline void set_running_frame(context f)
+{
+    current_cpu()->running_frame = f;
+}
+
+static inline void frame_push(context new)
```

Comment:
ok...I have to think about that, I didn't realize that the frames were chained. doesn't that already solve some of the special case handling that I was complaining about? I guess I'd been advocating that we just merge this stack of frames into the scheduling queue. does that really make sense?

Reply:
Well they weren't really chained - just one level. :) But, yes, if it was nestable to a few levels it would remove those specialized frames.

I guess it's confusing because in these transient cases we're not really trying to save a context to be restored at a later time, as with thread frames. We just want to retain one level of register save while possibly catching another - before popping our way back out. If we were to do our register saves to the stack as opposed to discrete frame records, this would be all automatic.

As far as I know, the only case where we want to hang on to a context to restore at some indeterminate later time is the thread frame...because, well, that's the only real notion of a thread that we have in the system.